In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv("winequalityN.csv")

In [3]:
df.head(10)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   
5  white            8.1              0.28         0.40             6.9   
6  white            6.2              0.32         0.16             7.0   
7  white            7.0              0.27         0.36            20.7   
8  white            6.3              0.30         0.34             1.6   
9  white            8.1              0.22         0.43             1.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   
5      0.050                 30.0                  97.0   0.9951  3.26   
6      0.045                 30.0                 136.0   0.9949  3.18   
7      0.045                 45.0                 170.0   1.0010  3.00   
8      0.049                 14.0                 132.0   0.9940  3.30   
9      0.044                 28.0                 129.0   0.9938  3.22   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6  
5       0.44     10.1        6  
6       0.47      9.6        6  
7       0.45      8.8        6  
8       0.49      9.5        6  
9       0.45     11.0        6

In [4]:
df.shape

(6497, 13)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   type                  6497 non-null   object 
 1   fixed acidity         6487 non-null   float64
 2   volatile acidity      6489 non-null   float64
 3   citric acid           6494 non-null   float64
 4   residual sugar        6495 non-null   float64
 5   chlorides             6495 non-null   float64
 6   free sulfur dioxide   6497 non-null   float64
 7   total sulfur dioxide  6497 non-null   float64
 8   density               6497 non-null   float64
 9   pH                    6488 non-null   float64
 10  sulphates             6493 non-null   float64
 11  alcohol               6497 non-null   float64
 12  quality               6497 non-null   int64  
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


In [6]:
df.isnull().sum()

type                     0
fixed acidity           10
volatile acidity         8
citric acid              3
residual sugar           2
chlorides                2
free sulfur dioxide      0
total sulfur dioxide     0
density                  0
pH                       9
sulphates                4
alcohol                  0
quality                  0
dtype: int64

In [7]:
df['fixed acidity'].fillna(df['fixed acidity'].mode()[0], inplace=True)
df['volatile acidity'].fillna(df['volatile acidity'].mode()[0], inplace=True)
df['citric acid'].fillna(df['citric acid'].mode()[0], inplace=True)
df['residual sugar'].fillna(df['residual sugar'].mode()[0], inplace=True)
df['chlorides'].fillna(df['chlorides'].mode()[0], inplace=True)
df['pH'].fillna(df['pH'].mode()[0], inplace=True)
df['sulphates'].fillna(df['sulphates'].mode()[0], inplace=True)

In [8]:
df.isnull().sum()

type                    0
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [9]:
X = df.drop(['type','quality'],axis=1)
y= df['quality']

In [10]:
#X=df.drop(['type','quality'],axis=1)
#y=df['quality']

**Splitting the Dataset**

In [11]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.30,random_state=42)

In [12]:
from sklearn.linear_model import LinearRegression
l = LinearRegression()
l.fit(X_train,y_train)
y_pred = l.predict(X_test)

In [13]:
from sklearn.metrics import mean_squared_error,r2_score
m_l = mean_squared_error(y_test,y_pred)
print("Mean Squared error of Linear model is ",m_l)

Mean Squared error of Linear model is  0.5072523112618679


In [14]:
r2_l = r2_score(y_test,y_pred)
r2_l

0.30528998963306986

**Standard scaling**

In [15]:
from sklearn.preprocessing import StandardScaler,MinMaxScaler
ss = StandardScaler()

In [16]:
X_train_ss = ss.fit_transform(X_train)

In [17]:
X_test_ss = ss.transform(X_test)

## **Creating the Linear Regression Model**

In [18]:
from sklearn.linear_model import LinearRegression
ln = LinearRegression()
ln.fit(X_train_ss,y_train)
y_pred_ln_ss = ln.predict(X_test_ss)

In [19]:
m_ln_ss = mean_squared_error(y_test,y_pred_ln_ss)
print("Mean Squared Error of Linear Model using Standard Scaling",m_ln_ss)

Mean Squared Error of Linear Model using Standard Scaling 0.5072523112618676


In [20]:
r2_ln_ss = r2_score(y_test,y_pred_ln_ss)
r2_ln_ss

0.3052899896330702

## **Creating the SVR model**

In [21]:
from sklearn.svm import SVR
svr = SVR()
svr.fit(X_train_ss,y_train)
y_pred_svr_ss = svr.predict(X_test_ss)

In [22]:
from sklearn.metrics import mean_squared_error,r2_score
m_svr_ss = mean_squared_error(y_test,y_pred_svr_ss)
r2_svr_ss = r2_score(y_test,y_pred_svr_ss)
print("Mean Squared Error of Support Vector Regressor is ",m_svr_ss)

Mean Squared Error of Support Vector Regressor is  0.44383219493862663


In [23]:
r2_svr_ss

0.3921473359481381

### Creating the SGD Regressor Model

In [24]:
from sklearn.linear_model import SGDRegressor
sgd = SGDRegressor()
sgd.fit(X_train_ss,y_train)
y_pred_sgd_ss = sgd.predict(X_test_ss)                   

In [25]:
m_sgd_ss = mean_squared_error(y_test,y_pred_sgd_ss)
print("Mean Squared Error of SGD Regressor is ",m_sgd_ss)

Mean Squared Error of SGD Regressor is  0.5073364521238721


In [26]:
r2_sgd_ss = r2_score(y_test,y_pred_sgd_ss)
r2_sgd_ss

0.30517475408299466

### **Creating the Ridge regression Model**

In [27]:
from sklearn.linear_model import Ridge
rid = Ridge()
rid.fit(X_train_ss,y_train)
y_pred_rid_ss = rid.predict(X_test_ss)

In [28]:
m_rid_ss = mean_squared_error(y_test,y_pred_rid_ss)
print("Mean Squared Error of Ridge regression is ",m_rid_ss)

Mean Squared Error of Ridge regression is  0.5072462720469213


In [29]:
r2_rid_ss = r2_score(y_test,y_pred_rid_ss)
r2_rid_ss

0.3052982606709439

### **Creating the Lasso Regression Model**

In [30]:
from sklearn.linear_model import Lasso
las = Lasso()
las.fit(X_train_ss,y_train)
y_pred_las_ss = las.predict(X_test_ss)

In [31]:
m_las_ss = mean_squared_error(y_test,y_pred_las_ss)
print("Mean Squared Error of Lasso Regression is ",m_las_ss)

Mean Squared Error of Lasso Regression is  0.7301694757431596


In [32]:
r2_las_ss = r2_score(y_test,y_pred_las_ss)
r2_las_ss

-7.358864998030512e-06

**Calculating the Root Mean Squared Error using Standard Scaling**

In [33]:
import numpy as np

In [34]:
print("RMSE of Linear Regression Model: ",np.sqrt(m_ln_ss))

print("RMSE of Support Vector Regressor model: ",np.sqrt(m_svr_ss))

print("RMSE of SGD regressor Model: ",np.sqrt(m_sgd_ss))

print("RMSE of Ridge Regression Model: ",np.sqrt(m_rid_ss))

print("RMSE of Lasso Regression Model: ",np.sqrt(m_las_ss))

RMSE of Linear Regression Model:  0.7122164778084453
RMSE of Support Vector Regressor model:  0.6662073212886711
RMSE of SGD regressor Model:  0.7122755450834123
RMSE of Ridge Regression Model:  0.7122122380631501
RMSE of Lasso Regression Model:  0.8544995469531623


**Finding out the best of Root Mean Squared Error**

_The Best value of Root Mean Squared error is 0.6662073212886711 and the model is Support Vector Regressor Model_